src/xenia/base/clock.cc

```diff
@@ -22,34 +33,87 @@ double guest_time_scalar_ = 1.0;
 uint64_t guest_tick_frequency_ = Clock::host_tick_frequency();
 // Base FILETIME of the guest system from app start.
 uint64_t guest_system_time_base_ = Clock::QueryHostSystemTime();
-// Combined time and frequency scalar (computed by RecomputeGuestTickScalar).
-double guest_tick_scalar_ = 1.0;
+// Combined time and frequency ratio between host and guest.
+// Split in numerator (first) and denominator (second).
+// Computed by RecomputeGuestTickScalar.
+std::pair<uint64_t, uint64_t> guest_tick_ratio_ = std::make_pair(1, 1);
+
 // Native guest ticks.
-thread_local uint64_t guest_tick_count_ = 0;
-// 100ns ticks, relative to guest_system_time_base_.
-thread_local uint64_t guest_time_filetime_ = 0;
+uint64_t last_guest_tick_count_ = 0;
 // Last sampled host tick count.
-thread_local uint64_t last_host_tick_count_ = Clock::QueryHostTickCount();
+uint64_t last_host_tick_count_ = Clock::QueryHostTickCount();
+// Mutex to ensure last_host_tick_count_ and last_guest_tick_count_ are in sync
+std::mutex tick_mutex_;
 
 void RecomputeGuestTickScalar() {
-  guest_tick_scalar_ = (guest_tick_frequency_ * guest_time_scalar_) /
-                       static_cast<double>(Clock::host_tick_frequency());
+  // Create a rational number with numerator (first) and denominator (second)
+  auto frac =
+      std::make_pair(guest_tick_frequency_, Clock::host_tick_frequency());
+  // Doing it this way ensures we don't mess up our frequency scaling and
+  // precisely controls the precision the guest_time_scalar_ can have.
+  if (guest_time_scalar_ > 1.0) {
+    frac.first *= static_cast<uint64_t>(guest_time_scalar_ * 10.0);
+    frac.second *= 10;
+  } else {
+    frac.first *= 10;
+    frac.second *= static_cast<uint64_t>(10.0 / guest_time_scalar_);
+  }
+  // Keep this a rational calculation and reduce the fraction
+  reduce_fraction(frac);
+  guest_tick_ratio_ = frac;
 }
 
-void UpdateGuestClock() {
-  uint64_t host_tick_count = Clock::QueryHostTickCount();
-  uint64_t host_tick_delta = host_tick_count > last_host_tick_count_
-                                 ? host_tick_count - last_host_tick_count_
-                                 : 0;
-  last_host_tick_count_ = host_tick_count;
-  uint64_t guest_tick_delta = uint64_t(host_tick_delta * guest_tick_scalar_);
-  guest_tick_count_ += guest_tick_delta;
-  guest_time_filetime_ += (guest_tick_delta * 10000000) / guest_tick_frequency_;
+// Update the guest timer for all threads.
+// Return a copy of the value so locking is reduced.
+uint64_t UpdateGuestClock() {
+  if (cvars::clock_no_scaling) {
+    // Nothing to update, calculate on the fly
+    auto host_tick_count = Clock::QueryHostTickCount();
+    return host_tick_count * guest_tick_ratio_.first / guest_tick_ratio_.second;
+  }
+
+  std::unique_lock<std::mutex> lock(tick_mutex_, std::defer_lock);
+
+  if (lock.try_lock()) {
+    // Translate host tick count to guest tick count.
+    uint64_t host_tick_count = Clock::QueryHostTickCount();
```

Comment:
Does this need to be under a lock? The others do since they're accessing member variables, but this can likely be executed concurrently and discarded upon "losing the race".